[FIXED JENKINS-39404] Restore symmetry by adding removeAction and removeActions methods to Actionable

stephenc opened the pull request:
See [JENKINS-39404](https://issues.jenkins-ci.org/browse/JENKINS-39404)

@jenkinsci/code-reviewers @reviewbybees

stephenc committed:
[FIXED JENKINS-39404] Restore symmetry by adding removeAction and removeActions methods to Actionable

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

stephenc committed:
[JENKINS-39404] Allow a more generic replaceActions

stephenc committed:
[JENKINS-39404] Damn generics... was unusable...

I would need to use <A super Action> which would allow Object.class

stephenc committed:
[JENKINS-39404] Minimize non-changes affecting write to disk requirement

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -204,24 +209,31 @@ public boolean removeActions(@Nonnull Class<? extends Action> clazz) {
     @SuppressWarnings({"ConstantConditions", "deprecation"})
     @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
-    public void replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
+    public boolean replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
```
As noted [here](http://stackoverflow.com/a/3589948/12916) this would not be binary-compatible even if the method were `final`, which it is not. You must restore the `void` return type. (Anyway you show no non-test callers using the return value and I doubt there would be any.)

jglick replied:
(looks like you reverted this later anyway)

stephenc replied:
This is a different *new* method replaceActionS

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -134,17 +141,99 @@ public void addAction(Action a) {
+     */
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
```
This should be typed as

```java
public <T extends Action> boolean replaceActions(@Nonnull Class<T> clazz, T a) {
```

stephenc replied:
No. The use case I have is replacing one type with a completely different implementation

stephenc replied:
Migrating form one set of actions to another

oleg-nenashev replied:
LGTM, complies with collection format

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -134,17 +141,99 @@ public void addAction(Action a) {
+    }
+
+    /**
+     * Remove an action.
```
Should clarify that `getAllActions()` could still contain it—just not `getActions()`.

jglick reviewed (changes requested):
You need to use the new methods from existing call sites which use `getActions` to remove things. Specifically I can find `Run.onLoad`, though I thought there were more (perhaps in plugins).

Also the `Computer.addAction` override should be deleted.

stephenc committed:
[JENKINS-39404] Make it easier for subclasses and address some code review comments

stephenc committed:
[JENKINS-39404] Clarify that you can replaceActions(Foo.class, Bar) where !(Bar instanceof Foo)

stephenc committed:
[JENKINS-39404] We cannot change these to return a boolean, but doesn't matter because they would always return `true`

stephenc commented:
@jglick I have addressed all your comments but not all in the way you may have intended

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -165,7 +178,8 @@ public void replaceAction(@Nonnull Action a) {
     /**
-     * Remove an action.
+     * Remove an action. Note calls to {@link #getAllActions()} that happen before calls to this method may not see the
+     * update.
```
Not just that. `getAllActions()` *after* this call may still see that action, if it was added by a transient factory.

jglick reviewed (approved):
:bee:

KostyaSha commented:
@stephenc the same for clouds would be good to see also. i.e. getter instead direct `cloud` field access.

oleg-nenashev commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -120,31 +133,124 @@
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
+        if (clazz == null) {
```
🐛 missing check/annotation for `a`, which is actually Nonnull in this code. Bug only since other methods check it

stephenc committed:
[JENKINS-39404] Introduce an `addOrReplaceAction` method with dirty return type

- We cannot modify `replaceAction` to return the dirty flag, in fact most callers probably don't care... but if you do care, we now provide `addOrReplaceAction`
- Clarify the javadocs to reflect that these methods are not going to affect actions added by transient action factories

oleg-nenashev commented:
my review is in progress

oleg-nenashev commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -121,30 +134,150 @@
+        for (Action a1 : current) {
+            if (!found && a.equals(a1)) {
+                found = true;
+            } else if (clazz.isInstance(a1) && !a.equals(a1)) {
```
Double `equals` call. We do not know how long it takes in implementations, so I would rather optimize by a local variable 🐜

stephenc commented:
@KostyaSha could you file a JIRA for the clouds one?

oleg-nenashev commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -121,30 +134,150 @@
+     * @since FIXME
+     */
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
```
🐛 I do not believe this method is thread-safe. We assume that actions get replaced by a single action, but Collection access and setter are not synchronized. Parallel calls of `replaceActions(clazz,b)` and `replaceActions(clazz,c)` may lead to the case when both `a` and `b` get to the collection. If `a` or `b` are instances of `clazz`, it will be an invalid behavior (If I read Javadoc correctly)

stephenc replied:
none of the existing methods are thread safe. parallel calls to `replaceAction(a)` and `replaceAction(b)` would run the same risk if they are the same class

oleg-nenashev replied:
Then we need to add proper annotation to all methods

stephenc replied:
Well `@NotThreadSafe` is for types not methods

oleg-nenashev replied:
Formally the class is not *completely* safe => NotThreadSafe, but I agree it's an overkill

oleg-nenashev commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -121,30 +134,150 @@
+            }
+        }
+        current.removeAll(old);
+        if (!found) {
```
🐛 I assume it should be `if (found)` if I read javadoc properly

stephenc replied:
Nope. We only add if we didn't find it, because if we found it then we didn't remove it and thus it is already there

stephenc replied:
Basically, we run through the actions, if we found the action in the list already, either you didn't override `equals` in which case it is == and therefore there is no point in removing the action just to add it again... or you did override `equals` in which case you have done that correctly and thus we would just be removing one equivalent object to replace with another.

oleg-nenashev reviewed (changes requested):
🐛 , mostly for `replaceActions` impl

stephenc committed:
[JENKINS-39404] Oleg wants to reduce the number of times we call `Action.equals()`

- cannot use a local variable as the optimization only applies while we have not found the action being added

stephenc committed:
[JENKINS-39404] Clarify the lack of transaction support and consequent effect on concurrent operations

stephenc commented:
@oleg-nenashev ping

oleg-nenashev commented:
🐝

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -72,16 +76,26 @@
+    /**
+     * We need to handle the initialization of the actions list in Actionable so that child classes that override
+     * getActions() for historical reasons do not have to override the manipulation methods: {@link #addAction(Action)},
+     * {@link #replaceAction(Action)}, {@link #removeAction(Action)}, etc.
```
I think this was an incompatible change: [JENKINS-39555](https://issues.jenkins-ci.org/browse/JENKINS-39555). `FlowNode.getActions()` was deliberately overridden to control storage.

svanoort replied:
I concur.   This was a *bad* core change.

jglick commented on core/src/main/java/hudson/model/Actionable.java:
```diff
@@ -72,16 +76,26 @@
+     * @return the CopyOnWriteArrayList of persisted actions.
+     */
+    private CopyOnWriteArrayList<Action> getOrCreateActions() {
+        if(actions == null) {
```
Yes, also looks like a mistake. Fixed in your amendment IIRC.
